Math/NumberTheory/EisensteinIntegers.hs

```diff
@@ -0,0 +1,345 @@
+-- @div@, or @quotRem@, which uses @quot@.)
+divHelper
+    :: (Integer -> Integer -> Integer)
+    -> EisensteinInteger
+    -> EisensteinInteger
+    -> (EisensteinInteger, EisensteinInteger)
+divHelper divide g h =
+    let nr :+ ni = g * conjugate h
+        denom = norm h
+        q = divide nr denom :+ divide ni denom
+        p = h * q
+    in (q, g - p)
+
+-- | Conjugate a Eisenstein integer.
+conjugate :: EisensteinInteger -> EisensteinInteger
+conjugate (a :+ b) = (a - b) :+ (-b)
+
+-- | The square of the magnitude of a Eisenstein integer.
+norm :: EisensteinInteger -> Integer
+norm (a :+ b) = a*a - a * b + b*b
+
+-- | Checks if a given @EisensteinInteger@ is prime. @EisensteinInteger@s
+-- whose norm is a prime congruent to @0@ or @1@ modulo 3 are prime.
+-- See <http://thekeep.eiu.edu/theses/2467 Bandara, Sarada, "An Exposition of the Eisenstein Integers" (2016)>,
+-- page 12.
+isPrime :: EisensteinInteger -> Bool
+isPrime e | e == 0                     = False
+          -- Special case, @1 - ω@ is the only Eisenstein prime with norm @3@,
+          --  and @abs (1 - ω) = 2 + ω@.
+          | a' == 2 && b' == 1         = True
+          | b' == 0 && a' `mod` 3 == 2 = Testing.isPrime a'
+          | nE `mod` 3 == 1            = Testing.isPrime nE
+          | otherwise = False
+  where nE       = norm e
+        a' :+ b' = abs e
+
+-- | Compute the GCD of two Eisenstein integers. The result is always
+-- in the first sextant.
+gcdE :: EisensteinInteger -> EisensteinInteger -> EisensteinInteger
+gcdE g h = gcdE' (abs g) (abs h)
+
+gcdE' :: EisensteinInteger -> EisensteinInteger -> EisensteinInteger
+gcdE' g h
+    | h == 0    = g -- done recursing
+    | otherwise = gcdE' h (abs (g `modE` h))
+
+-- | Find an Eisenstein integer whose norm is the given prime number
+-- in the form @3k + 1@ using a modification of the
+-- <http://www.ams.org/journals/mcom/1972-26-120/S0025-5718-1972-0314745-6/S0025-5718-1972-0314745-6.pdf Hermite-Serret algorithm>.
+--
+-- The maintainer <https://github.com/cartazio/arithmoi/pull/121#issuecomment-415010647 Andrew Lelechenko>
+-- derived the following:
+-- * Each prime of form @3n+1@ is actually of form @6k+1@.
+-- * One has @(z+3k)^2 ≡ z^2 + 6kz + 9k^2 ≡ z^2 + (6k+1)z - z + 9k^2 ≡ z^2 - z + 9k^2 (mod 6k+1)@.
+--
+-- * The goal is to solve @z^2 - z + 1 ≡ 0 (mod 6k+1)@. One has:
+-- @z^2 - z + 9k^2 ≡ 9k^2 - 1 (mod 6k+1)@
+-- @(z+3k)^2 ≡ 9k^2-1 (mod 6k+1)@
+-- @z+3k = sqrtMod(9k^2-1)@
+-- @z = sqrtMod(9k^2-1) - 3k@
+-- 
+-- * For example, let @p = 7@, then @k = 1@. Square root of @9*1^2-1 modulo 7@ is @1@.
+-- * And @z = 1 - 3*1 = -2 ≡ 5 (mod 7)@.
+-- * Truly, @norm (5 :+ 1) = 25 - 5 + 1 = 21 ≡ 0 (mod 7)@.
+findPrime :: Integer -> EisensteinInteger
+findPrime p = case Moduli.sqrtModMaybe (9*k*k - 1) (FieldCharacteristic (PrimeNat . integerToNatural $ p) 1) of
+    Nothing      -> error "findPrime: argument must be prime p = 6k + 1"
+    Just sqrtMod -> gcdE (p :+ 0) ((sqrtMod - 3 * k) :+ 1)
```

Comment:
Can we use the same approach as for Gaussian integers here? Complex GCD is very expensive.

Reply:
I forgot to leave a not to myself, but using the same process as Gaussian integers makes the process fail. Regarding the complexity of `gcdE`, I found [this](http://cs.au.dk/~gudmund/Documents/cubicres.pdf) link. Worth considering.

Reply:
@Bodigrim by the way, would you mind reading the paper I linked here and telling me how Lemma 1 can be practically implemented as an approximate norm function, please?

I did not understand the process behind it, only that it is possible and crucial for a fast Eisenstein `gcd`.

Reply:
Nevermind, I understood what was needed and the latest commit implements it.

Reply:
I didn't benchmark it directly, but benchmarks for `findPrime` have gotten considerably slower, and the only thing used by that function which changed was `gcdE`. This change might not be worth it after all.

I tried using the [same method](https://github.com/cartazio/arithmoi/blob/master/Math/NumberTheory/GaussianIntegers.hs#L178) that `Math.NumberTheory.GaussianIntegers.findPrime` uses, and as before, it does not work.